engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/IrPersistance.java

```diff
@@ -0,0 +1,391 @@
+  private IrPersistance() {}
+
+  @ServiceProvider(service = Persistance.class)
+  public static final class PersistIdentifiedLocation extends Persistance<IdentifiedLocation> {
+    public PersistIdentifiedLocation() {
+      super(IdentifiedLocation.class, false, 2);
+    }
+
+    @Override
+    protected void writeObject(IdentifiedLocation obj, Output out) throws IOException {
+      out.writeInline(Location.class, obj.location());
+    }
+
+    @Override
+    protected IdentifiedLocation readObject(Input in) throws IOException, ClassNotFoundException {
+      var obj = in.readInline(Location.class);
+      return new IdentifiedLocation((Location) obj, Option.empty());
+    }
+  }
+
+  @ServiceProvider(service = Persistance.class)
+  public static final class PersistScalaOption extends Persistance<Option> {
+    public PersistScalaOption() {
+      super(Option.class, true, 4431);
+    }
+
+    @Override
+    protected void writeObject(Option obj, Output out) throws IOException {
+      out.writeObject(obj.isEmpty() ? null : obj.get());
+    }
+
+    @Override
+    protected Option readObject(Input in) throws IOException, ClassNotFoundException {
+      var obj = in.readObject();
+      return Option.apply(obj);
+    }
+  }
+
+  @ServiceProvider(service = Persistance.class)
+  public static final class PersistScalaList extends Persistance<List> {
+    public PersistScalaList() {
+      super(List.class, true, 4432);
+    }
+
+    @Override
+    protected void writeObject(List list, Output out) throws IOException {
+      var size = list.size();
+      out.writeInt(size);
+      var l = list.reverse();
+      for (var i = 0; i < size; i++) {
+        out.writeObject(l.head());
+        l = (List) l.tail();
+      }
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    protected List readObject(Input in) throws IOException, ClassNotFoundException {
+      var size = in.readInt();
+      List list = scala.collection.immutable.Nil$.MODULE$;
+      for (var i = 0; i < size; i++) {
+        var elem = in.readObject();
+        list = scala.collection.immutable.$colon$colon$.MODULE$.apply(elem, list);
+      }
+      return list;
+    }
+  }
+
+  @ServiceProvider(service = Persistance.class)
+  public static final class PersistScalaSeq extends Persistance<Seq> {
+    public PersistScalaSeq() {
+      super(Seq.class, true, 4433);
+    }
+
+    @Override
+    protected void writeObject(Seq list, Output out) throws IOException {
+      var size = list.size();
+      out.writeInt(size);
+      for (var i = 0; i < size; i++) {
+        out.writeObject(list.apply(i));
+      }
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    protected Seq readObject(Input in) throws IOException, ClassNotFoundException {
+      var size = in.readInt();
+      Reference<?>[] arr = new Reference<?>[size];
+      for (var i = 0; i < size; i++) {
+        arr[i] = in.readReference(Object.class);
+      }
+      return new AbstractSeq() {
```

Comment:
Implementation of "lazy sequence" per @hubertp and @4e6 suggestions...

Reply:
nit: it would probably make sense to lift this into its own (private) class declaration taking `arr` in the constructor. It's a bit cumbersome to read inline

Reply:
Done in [b1ecbdf](https://github.com/enso-org/enso/pull/8207/commits/b1ecbdf4e2902754389a989490cf5b7f62918d0e)